File: 1647-minimum-deletions-to-make-character-frequencies-unique/1647-minimum-deletions-to-make-character-frequencies-unique.py
class Solution:
    def minDeletions(self, s: str) -> int:
        hashMap = dict()
        
        for ch in s:
            hashMap[ch] = hashMap.get(ch, 0) + 1
            
        maxHeap = []
        
        for key, value in hashMap.items():
            heappush(maxHeap, -value)
            
        count = 0
        while len(maxHeap):
            top = -heappop(maxHeap)
            
            if len(maxHeap) > 0 and -maxHeap[0] == top:
                if top != 1:
                    heappush(maxHeap, -(top - 1))
                count += 1
        return count
            

File: 1647-minimum-deletions-to-make-character-frequencies-unique/1647-minimum-deletions-to-make-character-frequencies-unique.cpp
class Solution {
public:
    int minDeletions(string s) {
        return daily(s);
    }
private:
    int daily(string s) {
        unordered_map<char, int> freqMap;
        priority_queue<int> freq;
        
        for(char ch: s)
            freqMap[ch]++;
        
        for(auto entry: freqMap)
            freq.push(entry.second);
        
        int count = 0;
        
        while(!freq.empty()) {
            int top = freq.top();
            freq.pop();
            
            if(!freq.empty() && freq.top() == top) {
                if(top != 1)
                    freq.push(top - 1);
                count++;
            }
        }
        return count;
    
    }
};